torchx/schedulers/ray/ray_driver.py

```diff
@@ -77,77 +102,189 @@ def load_actor_json(filename: str) -> List[RayActor]:
+            raise TimeoutError(
+                "Placement group creation timed out. Make sure "
+                "your cluster either has enough resources or use "
+                "an autoscaling cluster. Current resources "
+                "available: {}, resources requested by the "
+                "placement group: {}".format(
+                    ray.available_resources(), initial_group.bundle_specs
+                )
+            )
+        self.placement_groups.append(initial_group)
+        for i in range(1, len(replica_ix_of_pg) - 1):
+            self.placement_groups.append(
+                create_placement_group_async(
+                    self.replicas[replica_ix_of_pg[i] : replica_ix_of_pg[i + 1]]
+                )
+            )
 
-def create_command_actors(
-    actors: List[RayActor], pg: PlacementGroup
-) -> List[CommandActor]:
-    cmd_actors: List[CommandActor] = []
-    for i, replica in enumerate(actors):
-        # Environment variables for distributed training
+    def pop_actor_info(self, actor_id: str) -> ActorInfo:
+        """Remove and return the info of a dead command actor"""
+        return self.actor_info_of_id.pop(actor_id)
+
+    def create_and_schedule_actor(self, pg: PlacementGroup, replica: RayActor) -> None:
+        """create an command actor in the given placement group"""
+        # create the command actor
         actor = CommandActor.options(  # pyre-ignore[16]
             placement_group=pg,
             num_cpus=replica.num_cpus,
             num_gpus=replica.num_gpus,
         ).remote(replica.command, replica.env)
-        cmd_actors.append(actor)
-
-        if i == 0:
-            rank_0_address = "localhost"
-            rank_0_port = 0
-        else:
-            rank_0_address, rank_0_port = ray.get(
-                # pyre-ignore[16]
-                cmd_actors[0].get_actor_address_and_port.remote()
-            )
-        ray.get(actor.set_address_and_port.remote(rank_0_address, rank_0_port))
 
-    return cmd_actors
+        # get the actor id of the created actor
+        actor_id = actor._actor_id.hex()
+        # launch a task to check if the actor is scheduled
+        self.active_tasks.append(actor.schedule.remote(actor_id))
+        # save the actor info for recovering from node failures
+        self.actor_info_of_id[actor_id] = ActorInfo(
+            actor=actor,
+            pg=pg,
+            replica=replica,
+        )
+
+    def place_command_actors(self) -> None:
+        """Creating all command actors in all placement groups"""
+        # find the placement group index for a replica(actor's specification)
+        pg_ix_of_replica: List[int] = [
+            max(0, i - self.min_nnodes + 1)  # pyre-ignore
+            for i in range(len(self.replicas))
+        ]
+        # create the actors
+        for i in range(len(self.replicas)):
+            pg_ix = pg_ix_of_replica[i]
+            pg = self.placement_groups[pg_ix]  # find the created placement group
```

Comment:
I'm assuming if the placement_group hasn't started this just queues the actor?

Reply:
yes, it's just telling the ray cluster that I want to put a command actor in this placement group. And only after the placement group has been scheduled to an available node, the actors get created.